plugins/coordinator/plugin.go

```diff
@@ -99,42 +118,68 @@ func initCoordinator(bootstrap bool, startIndex uint32) (*coordinator.Coordinato
 		return nil, err
 	}
 
-	// initialize the selector
-	selector.Reset()
-
 	return coo, nil
 }
 
 func run(plugin *node.Plugin) {
-	// pass all new solid bundles to the selector
-	onBundleSolid := events.NewClosure(func(cachedBundle *tangle.CachedBundle) {
-		cachedBundle.ConsumeBundle(func(bndl *tangle.Bundle) { // bundle -1
 
-			if bndl.IsInvalidPastCone() || !bndl.IsValid() || !bndl.ValidStrictSemantics() {
-				// ignore invalid bundles or semantically invalid bundles or bundles with invalid past cone
-				return
+	// create a background worker that signals to issue new milestones
+	daemon.BackgroundWorker("Coordinator[MilestoneTicker]", func(shutdownSignal <-chan struct{}) {
+
+		timeutil.Ticker(func() {
+			// issue next milestone
+			select {
+			case nextMilestoneSignal <- struct{}{}:
+			default:
+				// do not block if already another signal is waiting
 			}
+		}, coo.GetInterval(), shutdownSignal)
 
-			selector.OnNewSolidBundle(bndl)
-		})
-	})
+	}, shutdown.PriorityCoordinator)
 
 	// create a background worker that issues milestones
 	daemon.BackgroundWorker("Coordinator", func(shutdownSignal <-chan struct{}) {
-		tanglePlugin.Events.BundleSolid.Attach(onBundleSolid)
-		defer tanglePlugin.Events.BundleSolid.Detach(onBundleSolid)
+		attachEvents()
 
-		// TODO: add some random jitter to make the ms intervals not predictable
-		timeutil.Ticker(func() {
-			err, criticalErr := coo.IssueNextCheckpointOrMilestone()
-			if criticalErr != nil {
-				log.Panic(criticalErr)
-			}
-			if err != nil {
-				log.Warn(err)
+		// bootstrap the network if not done yet
+		if criticalErr := coo.Bootstrap(); criticalErr != nil {
+			log.Panic(criticalErr)
+		}
+
+	coordinatorLoop:
+		for {
+			select {
+			case <-nextCheckpointSignal:
+				// check the thresholds again, because a new milestone could have been issued in the meantime
+				if tipCount, approveeCount := selector.GetStats(); (tipCount < maxTipsCount) && (approveeCount < maxApproveesCount) {
```

Comment:
Currently, a checkpoint is issued when the bitsets gets too large - this is absolutely crucial - and when the number of tips gets too large.
I am not so sure about the tip count, the heaviest tip selection is so fast that it would work for hundred thousands of tips. So what exactly is the motivation behind the tip limit? For performance it would make much more sense to limit the actual total runtime (like in the original version) instead of the number of tips. Also, if an attacker spams blow balls this would trigger many checkpoints which is actually not great.
If the purpose of that parameter is to trigger more checkpoints in very high TPS scenarios it would be favorable to use the weight of the heaviest tip as a threshold. This could be returned by the selector without computation.